x-pack/plugins/ingest_manager/common/types/rest_spec/epm.ts

```diff
@@ -71,6 +71,26 @@ export interface InstallPackageResponse {
   response: AssetReference[];
 }
 
+export interface UpgradePackageError {
+  name: string;
+  statusCode: number;
+  body: {
+    message: string;
+  };
+}
+
+export interface UpgradePackageInfo {
+  name: string;
+  newVersion: string;
+  // this will be null if no package was present before the upgrade (aka it was an install)
+  oldVersion: string | null;
+  assets: AssetReference[];
+}
+
+export interface UpgradePackagesResponse {
+  response: Array<UpgradePackageInfo | UpgradePackageError>;
```

Comment:
We have similar patterns elsewhere. Take a look at https://github.com/elastic/kibana/pull/77690/files#diff-f1874bde49bc5bca72d42584b63a5cf3R143-R147 & https://github.com/elastic/kibana/blob/71b9dedfc4eadd97db2b6b0dc038d0f640babf7b/x-pack/plugins/ingest_manager/common/types/rest_spec/package_policy.ts#L53-L57

I like `error: Error` if it's possible since it retains important info and can easily be handled/converted by others.

Reply:
Do you think it makes more sense to have the response of the bulk endpoint be something like:

```
export interface UpgradePackageInfo {
  name: string;
  newVersion?: string;
  // this will be null if no package was present before the upgrade (aka it was an install)
  oldVersion?: string | null;
  assets?: AssetReference[];
  statusCode: number;
  error?: Error;
}
```

If there's an error then, `newVersion, oldVersion, assets` will be undefined.

```
export interface UpgradePackagesResponse {
  response: UpgradePackageInfo[];
}
```